### ISSUE TITLE
pdns_control fails on Illumos / Solaris

### ISSUE THREAD

**Issue by UX-admin** [PowerDNS/pdns#2654]
Any command issued with pdns_control will emit the following message (which is in pdns/dynmessenger.cc):

Fatal error: Unable to set SO_RCVTIMEO option on socket: Option not supported by protocol

However, SO_RCVTIMEO is defined in /usr/include/sys/socket.h (this is from Solaris 10):

```
#define SO_SNDTIMEO     0x1005          /* send timeout */
#define SO_RCVTIMEO     0x1006          /* receive timeout */
```

Files affected, possibly more:
pdns/dynloader.cc
pdns/dynmessenger.hh
pdns/dynmessenger.cc

This is a show stopper for full integration with Solaris / Illumos / SmartOS / OmniOS / Tribblix service management facility: since SMF can be told to watch over configuration files in order to automatically signal a reload to a daemon when the content of a file (in this case, pdns.conf) changes, and if it tries to call pdns_control reload, the `svcadm refresh` will fail, and the service will be put into maintenance mode by SMF.

**Comment by rubenk** [comment]
We do include `<sys/socket.h>`. Perhaps Solaris doesn't support setting timeouts on unix sockets?

**Comment by rubenk** [comment]
Maybe it is this issue? https://www.illumos.org/issues/5898

**Comment by UX-admin** [comment]
If that is indeed the issue, it means that this has been fixed in illumos, but is still busted Solaris 10.

Since this did work in pdns_server 3.1, what do you think would be a viable workaround?

**Comment by rubenk** [comment]
I don't think it worked properly in pdns_server 3.1, unless I'm mistaken it just didn't have timeouts.

**Comment by cmouse** [comment]
Perhaps that shouldn't be a fatal error?

**Comment by UX-admin** [comment]
Whether or not it worked properly I cannot say; what I can say, by way of empirical testing, is that at the very least `pdns_control quit` did stop the pdns_server, and `pdns_control reload` appears to have reloaded the configuration. It is `pdns_control reload` part which is crucial to SMF, as that would be the refresh_method. Testing with pdns_server 3.1:

# pdns_server --version
Version: 3.1, compiled on Nov 28 2014, 15:26:21 with gcc version 4.7.1
# ps -eo pid,comm | fgrep pdns_server
#
# pdns_server
Sep 13 13:10:10 Reading random entropy from '/dev/urandom'
# ps -eo pid,comm | fgrep pdns_server
 2815 pdns_server
# pdns_control quit
Exiting
# ps -eo pid,comm | fgrep pdns_server
#
# pdns_server
Sep 13 13:12:49 Reading random entropy from '/dev/urandom'
# pdns_control reload
Ok
#

Based on this, I tend to agree with cmouse in that perhaps this should not be a fatal error. Is that possible to implement without compromising functionality or semantics?

**Comment by Habbie** [comment]
@UX-admin is this still a problem with modern pdns (4.1.0-rc2 or master) and a modern OS?

**Comment by Habbie** [comment]
ping

**Comment by UX-admin** [comment]
I'm still running 3.4.6, haven't tried a Solaris 10 build since. Is your own build on Solaris 10 successful?

**Comment by pieterlexis** [comment]
> Is your own build on Solaris 10 successful?

We don't build on Solaris

**Comment by UX-admin** [comment]
checking for Boost headers version >= 1.35.0... /opt/scfw/include/boost-1_38
checking for Boost's header version... sed: Missing newline at end of file standard input.
1_38
configure: error: invalid value: boost_major_version='sed: Missing newline at end of file standard input.
138'
configure: WARNING: cache variable boost_cv_lib_version contains a newline
gmake: *** No targets specified and no makefile found.  Stop.

Looks like https://github.com/PowerDNS/pdns/issues/2745 is still an issue, so I cannot build pdns on Solaris 10 without further fixes.

**Comment by pieterlexis** [comment]
I would recommend opening this issue (with a minimal reproduction) in https://github.com/tsuna/boost.m4.

**Comment by UX-admin** [comment]
"We don't build on Solaris"

...the effect of that is that pdns is undergoing regressions on both Solaris and illumos (a freeware, open source derivative of OpenSolaris).

There are several ways to mitigate this with relatively low overhead:

Solaris 10 in VirtualBox as a build and test server
or
building pdns-server and pdns-recursor in a SmartOS native zone.

Building and testing in a SmartOS native zone will at least make pdns-server and pdns-recursor current and attractive for all illumos-based users and will probably break it for Solaris 10, but it's a better option than the current state of affairs, which pretty much makes pdns completely unavailable on all Solaris and Solaris derivative platforms.

**Comment by UX-admin** [comment]
"I would recommend opening this issue (with a minimal reproduction) in https://github.com/tsuna/boost.m4."

This approach was already recommended initially when I reported this bug but I've refrained from engaging the Boost community because there has been no solution to any of the other Solaris bugs I've opened there. Prompted by that suboptimal experience several times before, I've opted to invest my scarce time elsewhere. Since this bug comes from upstream (Boost), I would recommend to apply the patch in your own codebase or contact them on behalf of all Solaris and illumos users, as they will perhaps give you greater priority over an anonymous user such as myself.

Long story short: I've gotten nowhere with the Boost developers in the past.

**Comment by pieterlexis** [comment]
> ...the effect of that is that pdns is undergoing regressions on both Solaris and illumos (a freeware, open source derivative of OpenSolaris).

To be frank, the only OS that makes commercial sense for us to build and test on is Linux. We welcome (and have merged) contributions from the FreeBSD, OpenBSD and other OS communities to improve building on those Operating Systems, but we do not have the manpower nor will to test each and every OS there is.

**Comment by ahupowerdns** [comment]
To clarify a little bit, we care about all kinds of platforms (including BSDs and OSX), but the less mainstream your platform is, the more work you'll have to do yourself to keep things working. This means finding fixes for things yourself, but we are available to provide hints and guidance on how to get things working. For Solaris and derivatives, it appears there just aren't any volunteers however, and we can't do the heavy lifting ourselves.

**Comment by UX-admin** [comment]
As someone who has very limited resources in the form of spare time, I can deeply symphatize with your sentiment and can definitely understand where you're coming from. However, if you set up a VirtualBox server with Solaris 10, you will have an operating system which sports reference implementations of any open protocol and specification imaginable, that is to say: if it builds correctly on Solaris, it will build everywhere because of that. Additionally, setting up a Solaris 10 machine inside of VirtualBox is fairly straightforward. If you want to flush out standard-compliance bugs, this is the operating system you want to build and test on, because it will trigger them.

**Comment by ahupowerdns** [comment]
I'm afraid we disagree. Good luck!

**Comment by UX-admin** [comment]
The problem is, I cannot run pdns on GNU/Linux if I want data integrity and correctness of operation, since basic semantics like fsync(2) do not work correctly and lots of that OS breaks and breaks often as far introducing non-backwards compatible changes. Solaris and illumos based operating systems remain the only practical alternative.

As it stands right now, fact is that I cannot verify what you ask because pdns cannot even get past the configure stage on Solaris.

**Comment by Habbie** [comment]
You are commenting on a closed ticket, which usually is a great way to accidentally get ignored.

Furthermore, your attitude is unproductive, unfriendly, and disrespectful. Example:

> As someone who has very limited resources in the form of spare time, I can deeply symphatize with your sentiment and can definitely understand where you're coming from. However,

No. There is no 'however' there. There is no 'now please spend a day on my obscure problem' after 'I understand that you have better things to do'.

I am closing comments on this ticket. Please reconsider how you ask (even almost demand) free help from open source projects, before contacting us again.
